src/main/groovy/au/org/ala/ecodata/reporting/CSProjectXlsExporter.groovy

```diff
@@ -158,22 +165,97 @@ class CSProjectXlsExporter extends ProjectExporter {
+
+        OutputModelProcessor processor = new OutputModelProcessor()
+
+        Set<String> uniqueOutputs = [] as HashSet<String>
+
+        List<Map> batchedActivities = []
         // BioCollect currently doesn't use form versioning so just get the latest version of the form.
         ActivityForm form = activityFormService.findActivityForm(projectActivity.pActivityFormName)
         if (activityIds == null || activityIds.isEmpty()) {
-            activities = activityService.findAllForProjectActivityId(projectActivity.projectActivityId)
-        } else {
-            activities = activityService.findAllForActivityIdsInProjectActivity(activityList, projectActivity.projectActivityId)
+            def id = new MinKey()
+
+            while (count == batchSize) {
+                batchedActivities = Activity.createCriteria().list([offset: 0, max: batchSize, readOnly: true, sort: 'id', order: "asc"]) {
+                    and {
+                        eq "projectActivityId", projectActivity.projectActivityId
+                        gt "id", id
+                    }
+                }
+
+                if (batchedActivities.size() > 0) {
+                    id = batchedActivities.last().id
+
+                    batchedActivities = batchedActivities.collect { activityService.toMap(it, []) }
+
+                    if (processed == 0) {
+                        batchedActivities.first().outputs.foreach { output ->
+                            Map model = form.getFormSection(output.name)?.template
+
+                            OutputMetadata outputModel = new OutputMetadata(model)
+                            Map outputConfig = getHeadersAndPropertiesForOutput(outputModel)
+                            headers.addAll(outputConfig.headers)
+                        }
+                    }
+
+                    sheet = generateSheet(batchedActivities, activityList, projectActivity, userId, userIsAlaAdmin, form, sheet, headers, processor, uniqueOutputs)
+                }
+
+                count = batchedActivities.size()
+                processed += batchedActivities.size()
+            }
         }
+        else {
+            def id = new MinKey()
+
+            while (count == batchSize) {
+                batchedActivities = Activity.createCriteria().list([offset: 0, max: batchSize, readOnly: true, sort: 'id', order: "asc"]) {
+                    eq "projectActivityId", projectActivity.projectActivityId
+                    'in' "activityId", activityList
+                    gt "id", id
+                }
 
-        if (activities && (activityIds == null || !activityIds.isEmpty())) {
-            List<String> headers = []
-            headers.addAll(surveyHeaders)
+                if (batchedActivities.size() > 0) {
+                    id = batchedActivities.last().id
+
+                    batchedActivities = batchedActivities.collect { activityService.toMap(it, []) }
+
+                    if (processed == 0) {
```

Comment:
It is possible that activities belong to different surveys in a project. Hence they can have multiple different survey forms. This code has to be modified to make sure those are added to header.

Reply:
Have a look at the previous code on how that was achieved.

Reply:
N/A